External idp impl

flo-renaud opened the pull request:
- External IdP schema: use octetstring syntax for client secret
- External Identity Provider plugin implementation
- External Idp: add XMLRPC tests

abbra commented on ipaserver/plugins/idp.py:
```diff
@@ -0,0 +1,304 @@
+    # List of pre-populated idp endpoints
+    # key = provider, value = tuple of (auth-uri, token-uri)
+    idp_providers = {
+        u'google': (
```
you don't need to use `u'...'` with Python 3 anymore, just use normal strings here and in other places, please.

abbra commented on ipaserver/plugins/idp.py:
```diff
@@ -0,0 +1,304 @@
+            'ipapermlocation': DN(container_dn, api.env.basedn),
+            'ipapermtargetfilter': {
+                '(objectclass=ipaidp)'},
+            'default_privileges': {'External Idp server Administrators'}
```
Please change `Idp` to `IdP` here.

abbra commented on ipaserver/plugins/idp.py:
```diff
@@ -0,0 +1,304 @@
+            'ipapermlocation': DN(container_dn, api.env.basedn),
+            'ipapermtargetfilter': {
+                '(objectclass=ipaidp)'},
+            'default_privileges': {'External Idp server Administrators'}
```
Please change `Idp` to `IdP` here.

abbra commented on ipaserver/plugins/idp.py:
```diff
@@ -0,0 +1,304 @@
+                'ipaidptokenendpoint', 'ipaidpclientid',
+                'ipaidpscope', 'ipaidpclientsecret',
+            },
+            'default_privileges': {'External Idp server Administrators'}
```
Please change `Idp` to `IdP` here.

abbra commented on ipaserver/plugins/idp.py:
```diff
@@ -0,0 +1,304 @@
+            'ipapermlocation': DN(container_dn, api.env.basedn),
+            'ipapermtargetfilter': {
+                '(objectclass=ipaidp)'},
+            'default_privileges': {'External Idp server Administrators'}
```
Please change `Idp` to `IdP` here.

abbra commented on ipatests/test_xmlrpc/test_idp_plugin.py:
```diff
@@ -0,0 +1,245 @@
+        command = idp_with_provider.make_create_command()
+        with raises_exact(errors.ValidationError(
+            name='provider',
+            error=u"must be one of 'google', 'github', 'microsoft-common', "
```
Same here -- no need to use `u"..."` for strings anymore

abbra reviewed (changes requested):
Thank you for the PR. It looks fine in general, couple comments inline. There are many places where you have used `u"..."` for unicode strings but this is not really needed with Python 3 anymore, could you please remove all the prefixes?
Also should we add 'admins' group to the default privilege?

flo-renaud commented:
Hi @abbra 
thanks for the review. I removed the unicode prefix (hope I didn't forget it anywhere) and fixed the uppercase in IdP.

> Also should we add 'admins' group to the default privilege?

By default admins can manage any entry thanks to the following ACI on the basedn:
```
aci: (targetattr != "userPassword || krbPrincipalKey || sambaLMPassword || sam
 baNTPassword || passwordHistory || krbMKey || krbPrincipalName || krbCanonica
 lName || krbPwdHistory || krbLastPwdChange || krbExtraData || krbLastSuccessf
 ulAuth || krbLastFailedAuth || ipaUniqueId || memberOf || enrolledBy || ipaNT
 Hash || ipaProtectedOperation")(version 3.0; acl "Admin can manage any entry"
 ; allow (all) groupdn = "ldap:///cn=admins,cn=groups,cn=accounts,dc=ipa,dc=te
 st";)

```

abbra commented:
Thanks. I think there is one place with `u'..'` left in tests. API.txt file has `u'..'` as well but those are generated and kept this way so they are OK.

flo-renaud commented:
I forgot to re-run makeaci and makeapi, it may fix some of the u'' issues. I'm on it...

abbra commented:
Right, it was not in the new code -- only in the context lines.

abbra commented:
Merged, thanks!
